### PR TITLE
libcontainer/intelrdt: support ClosID parameter

### PR DESCRIPTION
**Pull request by marquiz**
- Handle ClosID parameter of IntelRdt. Makes it possible to use
pre-configured classes/ClosIDs and avoid running out of available IDs
which easily happens with per-container classes.

- libcontainer/intelrdt: verify ClosID existence
Check that the ClosID directory pre-exists if no L3 or MB schema has
been specified. Conform with the following line from runtime-spec
(config-linux):

  *If closID is set, and neither of l3CacheSchema and memBwSchema are
  set, runtime MUST check if corresponding pre-configured directory
  closID is present in mounted resctrl. If such pre-configured directory
  closID exists, runtime MUST assign container to this closID and
  generate an error if directory does not exist.*

- Add a TODO note for verifying existing schemata against L3/MB
parameters.

- Run unit tests irrespective of the underlying system configuration, i.e.
even if RDT has not been enabled or is not supported. The tests do not
depend on real kernel interfaces.

### CODE REVIEW

File: libcontainer/intelrdt/intelrdt.go [hunk 1]
```diff
@@ -563,6 +562,14 @@ func (m *intelRdtManager) Apply(pid int) (err error) {
 	m.mu.Lock()
 	defer m.mu.Unlock()
 
+	if m.config.IntelRdt.ClosID != "" && m.config.IntelRdt.L3CacheSchema == "" && m.config.IntelRdt.MemBwSchema == "" {
+		// Check that the CLOS exists, i.e. it has been pre-configured to
+		// conform with the runtime spec
+		if _, err := os.Stat(path); err != nil {
+			return fmt.Errorf("unable to access %q (must be pre-created when l3CacheSchema and memBwSchema are empty): %v", path, err)
```
Comment (kolyshkin) [inline libcontainer/intelrdt/intelrdt.go:569]: 1. an error from os.Stat already contains path, no need to have it twice in the error message.
2. use `%w` for error.

File: libcontainer/intelrdt/intelrdt.go [hunk 2]
```diff
@@ -738,6 +745,12 @@ func (m *intelRdtManager) Set(container *configs.Config) error {
 		l3CacheSchema := container.IntelRdt.L3CacheSchema
 		memBwSchema := container.IntelRdt.MemBwSchema
 
+		// TODO: verify hat l3CacheSchema and/or memBwSchema match with
```
Comment (kolyshkin) [inline libcontainer/intelrdt/intelrdt.go:748]: typo: `s/ hat / that /`

**Comment by kolyshkin** [comment]
Looks good overall, added some notes.

Also, the first commit performs a nice refactoring (which I like), but it is kind of hard to review because it also introduces ClosID support. Can you maybe split it into (1) refactoring (2) adding ClosID?

File: libcontainer/intelrdt/intelrdt.go (hunk 2 above)
Comment (marquiz) [reply libcontainer/intelrdt/intelrdt.go]: Fixed

File: libcontainer/intelrdt/intelrdt.go (hunk 1 above)
Comment (marquiz) [reply libcontainer/intelrdt/intelrdt.go]: Valid point, changed now to `"clos dir not accessible (must be pre-created when l3CacheSchema and memBwSchema are empty): %w"`

**Comment by marquiz** [comment]
Thanks for the review @kolyshkin!

> Also, the first commit performs a nice refactoring (which I like), but it is kind of hard to review because it also introduces ClosID support. Can you maybe split it into (1) refactoring (2) adding ClosID?

Yeah, I see 😊 I now split the first commit into two

**Comment by marquiz** [comment]
BTW, related, I submitted a PR against runtime-spec in an attempt to make it easier read:
https://github.com/opencontainers/runtime-spec/pull/1104

**Comment by marquiz** [comment]
Ping @kolyshkin. Could you enable testing?

File: libcontainer/intelrdt/intelrdt.go [hunk 3]
```diff
@@ -738,6 +745,12 @@ func (m *intelRdtManager) Set(container *configs.Config) error {
 		l3CacheSchema := container.IntelRdt.L3CacheSchema
 		memBwSchema := container.IntelRdt.MemBwSchema
 
+		// TODO: verify that l3CacheSchema and/or memBwSchema match with
```
Comment (kolyshkin) [inline libcontainer/intelrdt/intelrdt.go:748]: nit: `s/match with/match the/`

File: libcontainer/intelrdt/intelrdt.go [hunk 4]
```diff
@@ -738,6 +745,12 @@ func (m *intelRdtManager) Set(container *configs.Config) error {
...
+		// existing schemata if ClosID has been specified. This is a more
+		// involved than reading the file and doing plain string comparison as
+		// the value written in does not necessarily match what gets read out
+		// (leading zeros, cache id ordering etc).
```
Comment (kolyshkin) [inline libcontainer/intelrdt/intelrdt.go:752]: Maybe it makes sense to relax the runtime spec and say "runtime MAY check" rather than "must check", because it if is a must and you're not doing it, you're not conformant.

File: libcontainer/intelrdt/intelrdt.go (hunk 3 above)
Comment (marquiz) [reply libcontainer/intelrdt/intelrdt.go]: Fixed

File: libcontainer/intelrdt/intelrdt.go (hunk 4 above)
Comment (marquiz) [reply libcontainer/intelrdt/intelrdt.go:752]: Might be, but I'm not entirely sure as it leads to inconsistency. User asking for `ClosID=foo,  L3Schema=bar` might get what she asks for or then something totally different.

Anyways, I think that should be debated in runtime-spec repo instead of runc.

File: libcontainer/configs/validate/validator.go [hunk 5]
```diff
@@ -200,19 +200,16 @@ func (v *ConfigValidator) intelrdt(config *configs.Config) error {
 			return errors.New("intelRdt is specified in config, but Intel RDT is not supported or enabled")
 		}
 
+		if config.IntelRdt.ClosID == "." || config.IntelRdt.ClosID == ".." || strings.ContainsRune(config.IntelRdt.ClosID, '/') {
```
Comment (kolyshkin) [inline libcontainer/configs/validate/validator.go:203]: I think `strings.ContainsRune` is overkill here, as we don't care about unicode, and `strings.Contains` should be sufficient (and probably faster).

File: libcontainer/configs/intelrdt.go [hunk 6]
```diff
@@ -1,6 +1,9 @@
 package configs
 
 type IntelRdt struct {
+	// The identity for RDT Class of Service
```
Comment (kolyshkin) [inline libcontainer/configs/intelrdt.go:4]: Perhaps you need to document it more, for example, say that
 * if not set, it equals the container ID;
 * if explicitly set, it is assumed that the corresponding resctrl directory is managed externally, and it is not removed by runc when he container is destroyed.

Comment (kolyshkin) [reply libcontainer/configs/intelrdt.go:4]: Or, if this is sufficiently documented in runtime-spec, please disregard the above comment.

Comment (marquiz) [reply libcontainer/configs/intelrdt.go:4]: K

File: libcontainer/configs/validate/validator.go (hunk 5 above)
Comment (marquiz) [reply libcontainer/configs/validate/validator.go]: Hmm, good point! I was thinking that `ContainsRune` would be better than substring matching but now that I look at the source code of the strings package that's not the case. Changed to `strings.Contains()`

**Comment by marquiz** [comment]
@kolyshkin anything I can do to advance this? The latest review comment(s) should be addressed, now

**Comment by marquiz** [comment]
ping @kolyshkin

**Comment by marquiz** [comment]
How to re-run tests??
/retest

**Comment by marquiz** [comment]
ping @kolyshkin. How to get this forward?

**Comment by cyphar** [comment]
We need to manually retrigger them, there isn't a bot to control them unfortunately.

**Comment by kolyshkin** [comment]
> you can retrigger it by force pushing a new commit

One other way is to close and reopen a PR.

**Comment by marquiz** [comment]
> (I've retriggered it.)

Thanks @cyphar! Everything's green now 😎

**Comment by marquiz** [comment]
ping @crosbymichael @mrunalp @dqminh @hqhq @cyphar @AkihiroSuda @kolyshkin

Any comments, advice?

File: libcontainer/specconv/spec_linux.go [hunk 7]
```diff
@@ -301,6 +301,9 @@ func CreateLibcontainerConfig(opts *CreateOpts) (*configs.Config, error) {
 		}
 		if spec.Linux.IntelRdt != nil {
 			config.IntelRdt = &configs.IntelRdt{}
+			if spec.Linux.IntelRdt.ClosID != "" {
+				config.IntelRdt.ClosID = spec.Linux.IntelRdt.ClosID
+			}
```
Comment (kolyshkin) [inline libcontainer/specconv/spec_linux.go:306]: So, this makes no sense, as both variables are strings, the default value is `""` and so `if` is totally redundant.

I just found this by looking into intelrtd code; will open a small fix PR soon.

Comment (kolyshkin) [reply libcontainer/specconv/spec_linux.go:306]: Resulted in not-so-small PR although I think it makes sense to have that before this one.

@marquiz PTAL  https://github.com/opencontainers/runc/pull/3041 🙏🏻

Comment (marquiz) [reply libcontainer/specconv/spec_linux.go]: Resolved by rebase

**Comment by marquiz** [comment]
Rebased

File: libcontainer/intelrdt/intelrdt.go [hunk 8]
```diff
@@ -579,7 +571,7 @@ func (m *intelRdtManager) Destroy() error {
 // restore the object later
 func (m *intelRdtManager) GetPath() string {
 	if m.path == "" {
-		m.path, _ = GetIntelRdtPath(m.id)
+		m.path, _ = m.getIntelRdtPath()
 	}
 	return m.path
 }
```
Comment (kolyshkin) [inline libcontainer/intelrdt/intelrdt.go:594]: Why not merge `getIntelRdtPath` into here, and use it everywhere?

File: libcontainer/intelrdt/intelrdt.go [hunk 9]
```diff
@@ -548,16 +535,21 @@ func (m *intelRdtManager) Apply(pid int) (err error) {
 	if m.config.IntelRdt == nil {
 		return nil
 	}
-	d, err := getIntelRdtData(m.config, pid)
+
+	path, err := m.getIntelRdtPath()
```
Comment (kolyshkin) [inline libcontainer/intelrdt/intelrdt.go:543]: Why not m.GetPath()?

Comment (kolyshkin) [reply libcontainer/intelrdt/intelrdt.go:543]: Ah, I see, getIntelRdtPath returns an error, while GetPath does not.

File: libcontainer/intelrdt/intelrdt.go (hunk 8 above)
Comment (kolyshkin) [reply libcontainer/intelrdt/intelrdt.go:594]: Nevermind.

File: libcontainer/configs/validate/validator.go [hunk 10]
```diff
@@ -205,19 +205,16 @@ func (v *ConfigValidator) intelrdt(config *configs.Config) error {
...
 		if !intelrdt.IsMBAEnabled() && config.IntelRdt.MemBwSchema != "" {
 			return errors.New("intelRdt.memBwSchema is specified in config, but Intel RDT/MBA is not enabled")
 		}
-
-		if intelrdt.IsCATEnabled() && config.IntelRdt.L3CacheSchema == "" {
-			return errors.New("Intel RDT/CAT is enabled and intelRdt is specified in config, but intelRdt.l3CacheSchema is empty")
-		}
-		if intelrdt.IsMBAEnabled() && config.IntelRdt.MemBwSchema == "" {
-			return errors.New("Intel RDT/MBA is enabled and intelRdt is specified in config, but intelRdt.memBwSchema is empty")
-		}
```
Comment (kolyshkin) [inline libcontainer/configs/validate/validator.go:220]: Took me some time to figure out why it was removed. It's a duplicate of the lines above!

Added by commit 27560ace2fcc0fdaa5af1 (PR #1632).

I'd put it into a separate commit, for the ease of review, but it's fine as it is.

**Comment by kolyshkin** [comment]
A linter warning that needs to be fixed; otherwise LGTM.

File: libcontainer/configs/validate/validator.go (hunk 10 above)
Comment (marquiz) [reply libcontainer/configs/validate/validator.go:220]: They were not duplicate but need to drop those checks in order to specify only ClosID and leave these fields empty. Clarified this in the commit message

**Comment by marquiz** [comment]
> A linter warning that needs to be fixed; otherwise LGTM.

Woops, fixed

**Review by kolyshkin — APPROVED** [review]
LGTM

**Comment by marquiz** [comment]
Rebased

**Comment by marquiz** [comment]
ping#2 @crosbymichael @mrunalp @dqminh @hqhq @cyphar @AkihiroSuda @kolyshkin

**Comment by marquiz** [comment]
ping @crosbymichael @mrunalp @dqminh @hqhq @cyphar @AkihiroSuda @kolyshkin ❗️

This starts to be critical as RDT support was merged in CRI-O and we really need the runc to support it, too.
https://github.com/cri-o/cri-o/pull/4830

**Review by kolyshkin — APPROVED** [review]
Still LGTM :)

@Akihiro @thaJeztah PTAL

File: libcontainer/intelrdt/intelrdt.go [hunk 11]
```diff
@@ -532,14 +518,18 @@ func IsMBAScEnabled() bool {
 }
 
 // Get the 'container_id' path in Intel RDT "resource control" filesystem
```
Comment (AkihiroSuda) [inline libcontainer/intelrdt/intelrdt.go:520]: Perhaps container_id -> ClosID ?

Comment (marquiz) [reply libcontainer/intelrdt/intelrdt.go]: Good catch, thanks. I addressed this in a separate patch, modifying some other code comments as well. I used the term *clos group*

**Review by AkihiroSuda — APPROVED** [review]

**Comment by marquiz** [comment]
Signed-off-by was missing from the last patch 😭

**Review by AkihiroSuda — APPROVED** [review]

**Comment by marquiz** [comment]
ping @kolyshkin

**Review by kolyshkin — APPROVED** [review]
LGTM
